storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3110,1754 @@ buf_block_t *fsp_get_latched_xdes_page(
+  if (cur_page_no == FIL_NULL)
+    return err;
+
+  fil_space_t *const space= fil_system.sys_space;
+  uint32_t extent_size= space_defrag->m_extent_size;
+
+  buf_block_t *block= fsp_get_latched_page(page_id_t{0, cur_page_no},
+                                           mtr, &err);
+  if (!block)
+    return err;
+
+  for (;;)
+  {
+    page_t *page= buf_block_get_frame(block);
+    uint32_t next_page_no= btr_page_get_next(page);
+    uint32_t cur_extent= (cur_page_no / extent_size) * extent_size;
+    uint32_t old_state= space_defrag->get_state(cur_extent);
+
+    if (old_state == XDES_FREE)
+    {
+fetch_next_page:
+      if (next_page_no == FIL_NULL)
+        break;
+      mtr->commit();
+      cur_page_no= next_page_no;
+
+      mtr->start();
+      mtr->x_lock_space(space);
+      block= fsp_get_latched_page(page_id_t{0, cur_page_no},
+                                  mtr, &err);
+      if (!block)
+        return err;
+      continue;
+    }
+
+    uint32_t new_extent= space_defrag->get_new_extent(cur_extent);
+    /* There is no need for extent to be changed */
+    if (new_extent == cur_extent)
+    {
+      if (level)
+      {
+        err= store_child_pages(block);
+        if (err) return err;
+      }
+      goto fetch_next_page;
+    }
+
+    buf_block_t *header_block=
+      fsp_get_latched_page(page_id_t{0, 0}, mtr, &err);
+    if (!header_block)
+      return err;
+
+    const fseg_header_t *seg_header= m_root->page.frame +
+      (level ? PAGE_HEADER + PAGE_BTR_SEG_TOP
+             : PAGE_HEADER + PAGE_BTR_SEG_LEAF);
+
+    buf_block_t *iblock;
+    fseg_inode_t *inode= fseg_inode_try_get(seg_header, 0, 0, mtr,
+                                            &iblock, &err);
+    if (!inode)
+      return err;
+
+    auto parent_it= m_parent_pages.find(cur_page_no);
+    if (parent_it == m_parent_pages.end())
+    {
+      err= DB_CORRUPTION;
+      return err;
+    }
+
+    uint32_t parent_page_no= uint32_t(parent_it->second);
+
+    uint32_t parent_offset= uint32_t(parent_it->second >> 32);
+
+    if (parent_offset > srv_page_size - FIL_PAGE_DATA_END)
+    {
+      err= DB_CORRUPTION;
+      return err;
+    }
+
+    pageOperator operation(header_block, iblock, inode, extent_size,
+                           cur_page_no, mtr);
+
+    associatedPages related_pages(block, mtr);
+    err= operation.prepare_new_page(new_extent, old_state == XDES_FSEG);
+
+    DBUG_EXECUTE_IF("allocation_prepare_fail", err= DB_CORRUPTION;);
+    if (err) return err;
+
+    err= related_pages.prepare(parent_page_no);
+    DBUG_EXECUTE_IF("relation_page_prepare_fail", err= DB_CORRUPTION;);
+
+    if (err) return err;
+
+    operation.complete_new_page();
+
+    /* After allocating the new page, try to prepare the steps
+    of page removal function. Because there is a possiblity that
+    last block in FSEG_NOT_FULL/FSP_FREE_FRAG/FSP_FREE last block
+    could've changed while allocating the new block. */
+    err= operation.prepare_old_page();
```

Comment:
Were we not supposed to prepare everything first, before completing anything? The `complete_new_page()` seems to be out of place here, or its name is confusing.

Reply:
We need to allocate the new page before preparing the step to remove the old page. Because there is 
a possibility that two operation can work on the same list and modify the last page in the same list.

Reply:
Do we actually need the separate prepare/complete logic, or would it suffice to create a wrapper function for `buf_page_get_gen()` that would allocate a copy for eventually restoring each page that may need to be modified? In that way, we could use the regular functions and at the very end, "roll back" the mini-transaction. This "roll back" function should also debug-assert that each block that was marked as modified in `mtr_t::m_memo` will actually be restored.